manpages/man/tiledb_dataset_generate.1
.TH tiledb_dataset_generate 1 "13 October 2015" "Version 0.1" "TileDB programs"
 
.SH NAME
tiledb_dataset_generate - generates synthetic data for a particular array schema
in a \fBCSV\fR or \fBbinary\fR form, \fBcompressed\fR or \fBuncompressed\fR.

.SH SYNOPSIS
.B tiledb_dataset_generate 
.BI "-A " "array_name " "-f " "filename " "-n " "cell_num " 
.br
[
.BI "-w " "workspace "
] [
.BI "-g " "group "
] [
.BI "-s " "seed "
] [
.BI "-F " "format "
] [
.BI "-l " "delimiter "
]

.SH DESCRIPTION
.B tiledb_dataset_generate
generates a sythetic dataset for a particular array schema in a \fBCSV\fR or
\fBbinary\fR format, \fBcompressed\fR or \fBuncompressed\fR. For the CSV format,
the user can specify any character \fBdelimiter\fR, including \fItab\fR. The
output dataset essentially consists of a collection of cells, whose format 
complies with the schema of the input array. This means that the given array
must already be defined. The coordinate values are drawn uniformly at random 
from their corresponding dimension domains defined in the array schema. The 
attribute values depend on their type. The real attribute values (i.e., 
\fBfloat32\fR and \fBfloat64\fR) are drawn uniformly at random from 
\fB[0.0,1.0]\fR. The \fBint32\fR and \fBint64\fR data values are drawn uniformly
at random from \fB[0,max()]\fR, where \fBmax()\fR is the maximum value of the 
corresponding type supported by the system that calls the function. Finally, the
\fBchar\fR values are drawn uniformly at random from ASCII domain 
\fB[45,126]\fR, i.e., all the characters between and including \fB'-'\fR and 
\fB'~'\fR.

The user can either specify explicitly the data format, or let the program 
derive it from the provided file name. For instance, if the user does not 
specify the format, then file name "my_file.bin" implies that the data will be 
stored in uncompressed binary format, whereas "my_file.csv.gz" implies that the 
data will be stored in GZIP-compressed CSV format. The CSV and binary formats 
are explained in detail in \fBtiledb_array_load\fR(1). Note that the generated data
do not contain deletions.

In addition, specifically or the case of CSV file, the use may specify a custom
delimiter. Finally, the user may specify a random seed, so that he/she can 
generate the same "random" dataset multiple times (e.g., for experiment
reproducibility purposes).

On error, it returns \fB-1\fR. Moreover, if it is compiled in \fBVERBOSE\fR 
mode, it prints comprehensive error messages on \fIstderr\fR. On success, it 
returns \fB0\fR. 

The following conventions apply to the \fBSYNOPSIS\fR section:

.TS
tab (@);
c lx .
\fBbold text\fR @ type exactly as shown
\fIitalic text\fR @ replace with appropriate argument
[\fB\-a \fIarg\fR]@ any or all options within [ ] are optional
.TE

.SH OPTIONS
.TP
.BI "-w" " workspace" "\fR, " \fB --workspace=\fIworkspace\fR  
The path to the workspace folder where the array is stored. If it is not 
provided, the current working directory is set as the workspace by default.

.TP
.BI "-g" " group" "\fR, " \fB --group=\fIgroup\fR  
This is a directory inside the workspace where the array is stored.
Note that any group path provided is converted into an absolute path 
\fBrelative to the workspace directory\fR, i.e., regarding all home ("\\~/"), 
current ("./") and root ("/") as the workspace directory. If the group is not 
provided, the workspace is set as the default group.

.TP
.BI "-A" " array_name" "\fR, " \fB --array-name=\fIarray_name\fR  
The name of the array for which the data are generated. The array
must be already defined. 

.TP
.BI "-f" " filename" "\fR, " \fB --filename=\fIfilename\fR  
The name of the output file that will store the generated data. If the user does
not specify explicitly the file format, the program derives it automaticalyy 
from the provided file name. For instance, if the user does not specify the 
format, then file name "my_file.bin" implies that the data will be stored in 
uncompressed binary format, whereas "my_file.csv.gz" implies that the data will 
be stored in GZIP-compressed CSV format.

.TP
.BI "-n" " cell_num" "\fR, " \fB --cell_num=\fIcell_num\fR  
The number of cells to be generated.

.TP
.BI "-s" " seed" "\fR, " \fB --seed=\fIseed\fR  
The seed that will be used internally by the random generator. This is
optional. If it is omitted, the default is the current time of day. 

.TP
.BI "-F" " format" "\fR, " \fB --format=\fIformat\fR  
It can be one of the following: \fBcsv\fR (CSV format), \fBcsv.gz\fR 
(GZIP-compressed CSV format), \fBbin\fR (binary format), \fBbin.gz\fR 
(GZIP-compressed binary format). If it is not provided, the program will derive
it from the file name as described above for option '\fB\-f\fR'.

.TP
.BI "-l" " delimiter" "\fR, " \fB --delimiter=\fIdelimiter\fR  
This is meaningful only for CSV format. It stands for the delimiter which 
separates the values in a CSV line in the CSV file. If not given, the default is
\fB','\fR. For using a tab as a delimiter, the user must specify it simply 
as \fBtab\fR.

.SH EXAMPLES
.TP
\fBtiledb_dataset_generate\fR \\ 
    \fB-w \fRmy_workspace/ \\
    \fB-g \fRmy_group/ \\
    \fB-A \fRmy_array \\
    \fB-f \fRmy_array \\
    \fB-F \fRcsv \\
    \fB-l \fR';' \\
    \fB-n \fR10000 \\
    \fB-s \fR 0

Creates CSV data for array \fImy_array\fR that is stored in directory
\fImy_workspace/my_group/my_array\fR. The output file is \fI"my_array"\fR.
The delimiter character for each CSV line/cell is \fI';'\fR.
The file will contain \fI10000\fR cells. A fixed seed \fB0\fR was used.

\fBtiledb_dataset_generate\fR \\ 
    \fB-w \fRmy_workspace/ \\
    \fB-g \fRmy_group/ \\
    \fB-A \fRmy_array \\
    \fB-f \fRmy_array \\
    \fB-F \fRcsv.gz \\
    \fB-l \fRtab \\
    \fB-n \fR10000 \\
    \fB-s \fR 0

Same as above but now the delimiter is \fItab\fR, and the CSV file is 
\fIGZIP-compressed\fR.

\fBtiledb_dataset_generate\fR \\ 
    \fB-w \fRmy_workspace/ \\
    \fB-g \fRmy_group/ \\
    \fB-A \fRmy_array \\
    \fB-f \fRmy_array.bin.gz \\
    \fB-n \fR10000

Same as the first example, but now the file format is automatically derived from
the file name as binary GZIP-compressed. Moreover, the seed used in the random 
generator is the \fIcurrent time\fR.

.SH EXIT STATUS
.TP 
\fB0\fR for success and \fB-1\fR for error.

.SH SEE ALSO
.BR "tiledb_array_define" "(1) " "tiledb_array_load" "(1) 

.SH AUTHOR
Stavros Papadopoulos <http://people.csail.mit.edu/stavrosp>, Copyright (c) 2015


tiledb_cmd/src/tiledb_dataset_generate.cc
/**
 * @file   tiledb_dataset_generate.cc
 * @author Stavros Papadopoulos <stavrosp@csail.mit.edu>
 *
 * @section LICENSE
 *
 * The MIT License
 * 
 * @copyright Copyright (c) 2015 Stavros Papadopoulos <stavrosp@csail.mit.edu>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 * 
 * @section DESCRIPTION
 *
 * Implements command "tiledb_dataset_generate".
 */

#include "csv_line.h"
#include "special_values.h"
#include "tiledb.h"
#include "utils.h"
#include <chrono>
#include <getopt.h>
#include <iostream>
#include <string>

#define PRINT_ERROR(x) std::cerr << "[TileDB] Error: " << x << ".\n"
#define PRINT(x) std::cout << "[TileDB] " << x << ".\n"

/** 
 * Parses the command options. It returns 0 on success. On error, it prints
 * a message on stderr and returns -1.
 */                                                            
int parse_options(
    int argc, 
    char** argv, 
    std::string& workspace,
    std::string& group,
    std::string& array_name, 
    std::string& filename,
    std::string& format,
    unsigned int& seed,
    int64_t& cell_num,
    char& delimiter) {
  // Initialization
  workspace = "";
  group = "";
  array_name = "";
  filename = "";
  format = "";
  std::string delimiter_str = "";
  std::string seed_str("");
  std::string cell_num_str("");

  // Auxiliary variable
  CSVLine temp;

  // *************** //
  // Parse options * //
  // *************** // 

  // Define long options
  static struct option long_options[] = 
  {
    {"array-name",1,0,'A'},
    {"filename",1,0,'f'},
    {"format",1,0,'F'},
    {"group",1,0,'g'},
    {"delimiter",1,0,'l'},
    {"cell-num",1,0,'n'},
    {"seed",1,0,'s'},
    {"workspace",1,0,'w'},
    {0,0,0,0},
  };

  // Define short options
  const char* short_options = "A:f:F:g:l:n:s:w:";
  // Get options
  int c;
  int option_num = 0;
  while((c=getopt_long(argc, argv, short_options, long_options, NULL)) >= 0) {
    ++option_num;
    switch(c) {
      case 'A':
        if(array_name != "") {
          PRINT_ERROR("More than one array names provided");
          return -1;
        }
        array_name = optarg;
        break;
      case 'f':
        if(filename != "") {
          PRINT_ERROR("More than one file names provided");
          return -1;
        }
        filename = optarg;
        break;
      case 'F':
        if(format != "") {
          PRINT_ERROR("More than one formats provided");
          return -1;
        }
        format = optarg;
        break;
      case 'g':
        if(group != "") {
          PRINT_ERROR("More than one groups provided");
          return -1;
        }
        group = optarg;
        break;
      case 'l':
        if(delimiter_str != "") {
          PRINT_ERROR("More than one delimiters provided");
          return -1;
        }
        delimiter_str = optarg;
        break;
      case 'n':
        if(cell_num_str != "") {
          PRINT_ERROR("More than one numbers of cells provided");
          return -1;
        }
        cell_num_str = optarg;
        break;
      case 's':
        if(seed_str != "") {
          PRINT_ERROR("More than seeds provided");
          return -1;
        }
        seed_str = optarg;
        break;
      case 'w':
        if(workspace != "") {
          PRINT_ERROR("More than one workspaces provided");
          return -1;
        }
        workspace = optarg;
        break; 
      default:
        return -1;
    }
  }

  // ****************** //
  // Check correctness  //
  // ****************** //

  // Check number of arguments
  if((argc-1) != 2*option_num) {
    PRINT_ERROR("Arguments-options mismatch");
    return -1;
  }
  // Check if correct arguments have been ginen
  //----- array name ----- //
  if(array_name == "") {
    PRINT_ERROR("Array name not provided");
    return -1;
  }
  temp << array_name;
  if(temp.val_num() > 1) {
    PRINT_ERROR("More than one array names provided");
    return -1;
  }
  // ----- workspace ----- //
  temp.clear();
  temp << workspace;
  if(temp.val_num() > 1) {
    PRINT_ERROR("More than one workspaces provided");
    return -1;
  }
  // ----- group ----- //
  temp.clear();
  temp << group;
  if(temp.val_num() > 1) {
    PRINT_ERROR("More than one groups provided");
    return -1;
  }
  // ----- filename ----- //
  if(filename == "") {
    PRINT_ERROR("File name not provided");
    return -1;
  }
  temp.clear();
  temp << filename;
  if(temp.val_num() > 1) {
    PRINT_ERROR("More than one file names provided");
    return -1;
  }
  // ----- cell number ----- //
  if(cell_num_str == "") {
    PRINT_ERROR("Number of cells not provided");
    return -1;
  }
  temp.clear();
  temp << cell_num_str;
  if(temp.val_num() > 1) {
    PRINT_ERROR("More than one numbers of cells provided");
    return -1;
  }
  if(!is_non_negative_integer(cell_num_str.c_str())) {
    PRINT_ERROR("The number of cells must be a non-negative integer");
    return -1;
  } else {
    sscanf(cell_num_str.c_str(), "%lld", &cell_num); 
  }
  // ----- seed ----- //
  temp.clear();
  temp << seed_str;
  if(temp.val_num() > 1) {
    PRINT_ERROR("More than one seeds provided");
    return -1;
  }
  if(seed_str == "") {
    seed = std::chrono::system_clock::now().time_since_epoch().count();
  } else if(!is_non_negative_integer(seed_str.c_str())) {
    PRINT_ERROR("The seed must be a non-negative integer");
    return -1;
  } else {
    sscanf(seed_str.c_str(), "%u", &seed); 
  }
  // ----- format ----- //
  if(format == "") { // Format not given, derive it from filename
    if(ends_with(filename, ".csv")) {
      format = "csv"; 
    } else if(ends_with(filename, ".csv.gz")) {
      format = "csv.gz"; 
    } else if(ends_with(filename, ".bin")) {
      format = "bin"; 
    } else if(ends_with(filename, ".bin.gz")) {
      format = "bin.gz"; 
    } else {
      PRINT_ERROR("Cannot derive file format.");
      return -1;
    }
  } else { // Format is given, check correctness
    if(format != "csv" && format != "csv.gz" &&
       format != "bin" && format != "bin.gz") {
      PRINT_ERROR("Invalid file format");
      return -1;
    }
  }
  // ----- delimiter ----- //
  if(delimiter_str == "") {
    delimiter = ',';
  } else if(ends_with(format, "bin") || ends_with(format, "bin.gz")) {
    PRINT_ERROR("The delimiter is meaningless for binary format");
    return -1;
  } else if(delimiter_str == "tab") {
      delimiter = '\t';
  } else if(delimiter_str.size() == 1) {
      delimiter = delimiter_str[0];
  } else {
      PRINT_ERROR("Invalid delimiter");
      return -1;
  }
 
  return 0;
}

int main(int argc, char** argv) {
  // Parse command line
  std::string workspace;
  std::string group;
  std::string array_name;
  std::string filename;
  std::string format;
  unsigned int seed;
  int64_t cell_num;
  char delimiter;

  if(parse_options(
         argc, argv, 
         workspace, group, array_name, 
         filename, format, seed, cell_num, delimiter)) {
    PRINT_ERROR("Program failed");
    return -1;
  }

  // Initialize TileDB
  TileDB_CTX* tiledb_ctx;
  if(tiledb_ctx_init(tiledb_ctx))
    return -1;

  // Generate data
  if(tiledb_dataset_generate(
       tiledb_ctx, workspace.c_str(), group.c_str(),
       array_name.c_str(), filename.c_str(), format.c_str(),
       seed, cell_num, delimiter)) {
    tiledb_ctx_finalize(tiledb_ctx);
    PRINT_ERROR("Program failed");
    return -1;
  }

  // Finalize TileDB
  if(tiledb_ctx_finalize(tiledb_ctx))
    return -1;

  // Success
  PRINT("Program executed successfully");
  
  return 0;
}
